libbeat/publisher/queue/diskqueue/queue.go

```diff
@@ -0,0 +1,296 @@
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package diskqueue
+
+import (
+	"errors"
+	"fmt"
+	"os"
+	"sync"
+
+	"github.com/elastic/beats/v7/libbeat/common"
+	"github.com/elastic/beats/v7/libbeat/feature"
+	"github.com/elastic/beats/v7/libbeat/logp"
+	"github.com/elastic/beats/v7/libbeat/publisher/queue"
+)
+
+// diskQueue is the internal type representing a disk-based implementation
+// of queue.Queue.
+type diskQueue struct {
+	logger   *logp.Logger
+	settings Settings
+
+	// Metadata related to the segment files.
+	segments *diskQueueSegments
+
+	// Metadata related to consumer acks / positions of the oldest remaining
+	// frame.
+	acks *diskQueueACKs
+
+	// The queue's helper loops, each of which is run in its own goroutine.
+	readerLoop  *readerLoop
+	writerLoop  *writerLoop
+	deleterLoop *deleterLoop
+
+	// Wait group for shutdown of the goroutines associated with this queue:
+	// reader loop, writer loop, deleter loop, and core loop (diskQueue.run()).
+	waitGroup *sync.WaitGroup
+
+	// The API channels used by diskQueueProducer to send write / cancel calls.
+	producerWriteRequestChan  chan producerWriteRequest
+	producerCancelRequestChan chan producerCancelRequest
+
+	// writing is true if a writeRequest is currently being processed by the
+	// writer loop, false otherwise.
+	writing bool
+
+	// reading is true if the reader loop is processing a readBlock, false
+	// otherwise.
+	reading bool
+
+	// deleting is true if the segment-deletion loop is processing a deletion
+	// request, false otherwise.
+	deleting bool
+
+	// pendingFrames is a list of all incoming data frames that have been
+	// accepted by the queue and are waiting to be sent to the writer loop.
+	// Segment ids in this list always appear in sorted order, even between
+	// requests (that is, a frame added to this list always has segment id
+	// at least as high as every previous frame that has ever been added).
+	pendingFrames []segmentedFrame
+
+	// blockedProducers is a list of all producer write requests that are
+	// waiting for free space in the queue.
+	blockedProducers []producerWriteRequest
+
+	// This value represents the oldest frame ID for a segment that has not
+	// yet been moved to the acked list. It is used to detect when the oldest
+	// outstanding segment has been fully acknowledged by the consumer.
+	oldestFrameID frameID
+
+	// This lock must be held to read and write acked and ackedUpTo.
+	ackLock sync.Mutex
+
+	// The lowest frame id that has not yet been acknowledged.
+	ackedUpTo frameID
+
+	// A map of all acked indices that are above ackedUpTo (and thus
+	// can't yet be acknowledged as a continuous block).
+	// TODO: do this better.
+	acked map[frameID]bool
```

Comment:
> I’m not aware of all technical requirements or assumptions, but I wonder what would be implications of merging loop logic (e.g. writer and deleter) or even having a single loop.

The short answer is, merging loops would make things more complex. I tried to find ways to have fewer, but the operations (read / write / delete) are logically independent (none _needs_ to know about the others), and trying to merge them ends up being a mess trying to keep any one operation from blocking the others (e.g.: just because we get an error deleting an old segment file doesn't mean we should stop writing to the current segment, but if they share a goroutine this becomes difficult.) The best we could hope for would be more or less concatenating their code into a single file, but in reality it would be worse because of the sequencing dependencies it would introduce.

A writeup of the implementation design and its rationale is pending, but I'm also happy to zoom if you want to talk through these tradeoffs in the meantime :-)

> 1. It would be nice to cover new source code with test cases. You can also prepare sample state files and put them somewhere in the repo for better understanding of the file format.

Agreed 😅 An important goal of this design was its testability: all of the logic / state changes can be tested in isolation without live helper loops or disk I/O. But, the tests themselves are still pending...

> 2. What do you think about package-separating store file representation (serialization, deserialization, file format, frames) from operating logic.

Hmm -- I need to think about it more, but this makes sense. It would presumably be a package that exports the ability to read or write a segment from a given stream with helpers to [de]serialize `producer.Event`.

I can't think of any essential obstacles to this right now? I will try it next time I do a file organization pass.

> 3. Separate facade (implementation of queue API) from internal loops.

How do you mean -- they are "separated" right now into different source files and all internal details are private, are you suggesting an additional encapsulation layer somewhere?

> 4. Is the binary format recommended for persistent state due to performance reasons or is it feasible to use a human readable one at least for some part of it (to make debugging process easier)?

The binary format is what was agreed on a while ago before starting the implementation [I will ping you offline with the doc]. We could probably make it fully human readable with relatively minor performance loss, though it would make the reader / writer code more complicated.

For what it's worth, the current version is "nearly human readable" -- the data in the segment files is essentially a sequence of `[four byte length] [json blob] [four byte checksum] [trailing four byte length]` for each event, so the content of each frame is already readable, and binary length / checksum is about the simplest "non-human-readable" data to work with.

Making it completely human readable would probably mean attaching the checksum as a field and wrapping each frame in a json object. This would require the reader to be more clever, since it would need to parse / unescape the json as it reads to detect when each event is complete, handle bounds checking differently, etc. It would also require a bit more care in serialization, since frame size would no longer be data-blob-size-plus-constant, and similarly we'd need to be careful about the precise string on which the checksum is computed -- we'd need to do at least some manual json mangling to separate the checksum from the "real" frame data before using a generic parser.

All that said, this has some appeal now that there's a nearly-finished implementation and we can measure the performance implications... For now I'd like to keep it as is, since I want to finish all the pieces before adding potential complications, but if time allows I'd like to come back later to see if a json-only approach is feasible.

...Unless you were talking about the "state file" rather than the segments, in which case, yes that certainly can be human readable and I'm probably going to change that soon :-)

> 5. I saw a couple of panics in the PR. I’m not sure if it’s standard for libbeat or temporary workaround?

The only panics I know of right now are during shutdown, basically because producer cancellation / queue shutdown is only half-written -- if you see any panics _before_ shutdown I'd like to know about it though.

> 6. I wonder what are the differences between diskqueue and memqueue. It should be natural that the major difference is the place where data are stored. It would be nice to let one implementation depend on the other (of course not necessary as part of this PR).

Well... that's a really big difference :-) In memory we can just use a linked list to live event data, so there's no [de]serialization, and no particular action to take to move things in / out of the enqueued state, whereas in the disk queue almost all of the logic is centered around getting things to and from the disk in a timely way. Honestly I can't see any way to productively share code between them since none of the constraints the disk queue was designed around apply to the memory queue, though I could reexamine the memory queue at some point with this in mind.

> 7. Can autoRetryReader block permanently the diskqueue from processing (if error occurs)?

No -- it only retries on immediately-retriable errors `EAGAIN` and `EINTR` which by definition should be ephemeral (i.e. if we keep retrying they are guaranteed to eventually succeed or return a more permanent error).

Reply:
Ah, no, these were formerly used but their logic was just moved into `diskQueueACKs` and the old fields weren't cleaned up yet :-)